Additions to tests for literal inference

georgemarrows opened the pull request:
@andymcn - does this look like the kind of thing you want for #570 ?

Future extensions that I can think of
- other numeric types
- literals passed to functions
- literals in tuples
- generics

andymcn commented:
That's exactly the kind of thing I was thinking of.

A few points:
* Any function containing gtest ASSERTs must be called from within the `DO()` macro. This is due to how gtest propagates test failures. So everywhere you call `check_type_for_simple_let()` you should actually be doing `DO(check_type_for_simple_let())`.
* The function `check_type_for_simple_let()` might be too specific to be useful in many of the tests, ie any that aren't based on an assignment. It might be better to either make it more general or rename it something to do with "assign".
* Inferring to a union containing a duplicate type (LetUnionSameType) should indeed work. It's legal to assign a `U8` to something of type `(U8 | U8)`, so that's how we should infer it.
* The error "Multiple possible types for literal" for `(U8 & U32)` is indeed wrong. Hopefully I've just used the wrong error message somewhere.

georgemarrows commented:
Thanks for the detailed feedback!

- I've added DO and improved the name of the `check` function (I think).
- I've also added a test looking at generics, which involved a good bit of poking about in literal.c to (2%) understand what was going on. The changes in util.cc are a result.
- Is the happy generic test ok? I wasn't sure if it was useful to check that the inferred type is 'A'.
- Git-wise, do you want the fix-up commits squashing and do you want a merge from master or a rebase?

SeanTAllen commented:
Squashing would be great @georgemarrows.

Thanks!
